File: ddd-start2-study/src/main/java/com/example/ddd/board/application/BoardService.java
package com.example.ddd.board.application;

import com.example.ddd.board.domain.Board;
import com.example.ddd.board.domain.BoardRepository;
import com.example.ddd.board.domain.Comment;
import com.example.ddd.board.domain.CommentRepository;
import com.example.ddd.board.presentation.CreateBoardRequest;
import com.example.ddd.board.presentation.CreateCommentRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional
public class BoardService {

    private final BoardRepository boardRepository;
    private final CommentRepository commentRepository;

    public void createBoard(CreateBoardRequest request) {
        boardRepository.save(Board.builder()
                .title(request.getTitle())
                .contents(request.getContents())
                .build());
    }

    public void createComment(Long boardId, CreateCommentRequest request) {
        Board board = boardRepository.findById(boardId).orElseThrow();
        commentRepository.save(Comment.builder()
                .board(board)
                .comments(request.getComments())
                .build());
    }

    public void deleteBoard(Long boardId) {
        boardRepository.deleteById(boardId);
    }

    public void deleteComment(Long boardId, Long commentId) {
        Board board = boardRepository.findWithCommentsById(boardId).orElseThrow();
        board.deleteCommentById(commentId);
    }
}

File: board-sample/backend/board/src/main/kotlin/com/example/board/service/BoardService.kt
package com.example.board.service

import com.example.board.request.BoardCommand
import com.example.board.repository.BoardRepository
import com.example.board.response.BoardView
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.lang.IllegalArgumentException

@Service
class BoardService(
    private val boardRepository: BoardRepository
) {

    @Transactional(readOnly = true)
    fun findPage(pageable: Pageable): Page<BoardView> = boardRepository.findAll(pageable)
        .map { BoardView.of(it) }

    @Transactional(readOnly = true)
    fun findById(id: Long): BoardView? = boardRepository.findByIdOrNull(id)
        ?.run(BoardView.Companion::of)

    @Transactional(readOnly = false)
    fun modify(id: Long, command: BoardCommand): BoardView = boardRepository.findByIdOrNull(id)
        ?.run { this.modify(command.toCommand()) }
        ?.run(BoardView.Companion::of)
        ?: throw IllegalArgumentException("찾는 게시글이 없습니다.")

    @Transactional(readOnly = false)
    fun delete(id: Long) {
        boardRepository.deleteById(id)
    }
}